### PR TITLE
[package publishing] Add new publishing flow to cli and to smoke tests

### PR DESCRIPTION
**Pull request by wrwg**
### Description

This adds the new publishing flow to the CLI, reachable via the command `move publish --new-flow`. This specifically means deriving the `PackageMetadata` which we upload to chain programmatically from the Move package. This functionality, as well as basic package build functionality, has been factored out into a new module `move_tool/built_package.rs` (happily accepting better names).  

The new `BuiltPackage` is intended to become the general way how to build Move code programmatically throughout aptos-core. This PR also applies it to the smoke tests, and adding test cases for the new flow there as well. Next we can use `BuiltPackage` for e2e-tests in aptos-move to switch them over from MoveIR!

Notice that by default, `code::publish_package` stays disabled. One need to have a binary for which aptos_vm was built with the `fuzzing` feature to execute the new flow.

Other changes:

- Simplifies package metadata (for now) to just store the textual version of the Move.toml. We will see how much we need to unpack from this (e.g. dependencies).

### Test Plan

Smoke tests. Functional testing of upgrades happens in aptos-move/e2e-testsuite.

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2298)

### CODE REVIEW

File: crates/aptos/debug-move-example/Move.toml [hunk 1]
```diff
@@ -7,4 +7,3 @@ DebugDemo = "0x1"
 
 [dependencies]
 AptosFramework = { local = "../../../aptos-move/framework/aptos-framework" }
-MoveNursery = {local = "../../../aptos-move/framework/move-stdlib/nursery"}
```
Comment (movekevin) [inline crates/aptos/debug-move-example/Move.toml:10]: Why delete this?

**Review by movekevin — APPROVED** [review]

File: crates/aptos/debug-move-example/Move.toml (hunk 1 above)
Comment (wrwg) [reply crates/aptos/debug-move-example/Move.toml:10]: The nursery had been removed from our repo a while ago, so this wasn't compiling any longer. Drive by effect as I used it as a test case. Nothing you should be concerned about.

**Comment by wrwg** [comment]
Thank you so much @movekevin for approving! Waiting for more reviewers for merging.

File: crates/aptos/src/move_tool/mod.rs [hunk 2]
```diff
@@ -289,31 +301,42 @@ impl CliCommand<TransactionSummary> for PublishPackage {
...
+                .await
+                .map(TransactionSummary::from)
+        } else {
+            // Send the compiled module and metadata using the code::publish_package_txn.
+            let metadata = package.extract_metadata(
+                package.name().to_owned(),
+                upgrade_policy.unwrap_or_else(UpgradePolicy::compat),
+            )?;
+            let payload = aptos_transaction_builder::aptos_stdlib::code_publish_package_txn(
+                bcs::to_bytes(&metadata).expect("PackageMetadata has BCS"),
```
Comment (gregnazario) [inline crates/aptos/src/move_tool/mod.rs:332]: Let's not do any expects, and just do `?` to convert it to errors.  Avoiding unwraps and expects in the CLI space as it doesn't end up bubbling up to a parsable format

File: testsuite/smoke-test/src/aptos/move_test_helpers.rs [hunk 3]
```diff
@@ -4,43 +4,50 @@
...
+    ctx: &mut AptosContext<'_>,
+    move_dir: PathBuf,
+    package_name: impl Into<String>,
+    upgrade_policy: UpgradePolicy,
+) -> Result<TransactionFactory> {
+    let package = BuiltPackage::build(MovePackageDir::new(move_dir), true, true)?;
+    let blobs = package.extract_code();
+    let metadata = package.extract_metadata(package_name.into(), upgrade_policy)?;
+    let payload = aptos_transaction_builder::aptos_stdlib::code_publish_package_txn(
+        bcs::to_bytes(&metadata).expect("PackageMetadata has BCS"),
```
Comment (gregnazario) [inline testsuite/smoke-test/src/aptos/move_test_helpers.rs:44]: similarly with the ?

**Review by gregnazario — APPROVED** [review]
Looks good!  Thanks for doing this, just some minor comments about some wording

File: crates/aptos/src/common/types.rs [hunk 4]
```diff
@@ -689,7 +689,15 @@ pub struct MovePackageDir {
...
-    pub fn get_package_dir(&self) -> CliTypedResult<PathBuf> {
+    pub fn new(package_dir: PathBuf) -> Self {
+        Self {
+            package_dir: Some(package_dir),
+            output_dir: None,
+            named_addresses: Default::default(),
+        }
+    }
+
+    pub fn get_package_path(&self) -> CliTypedResult<PathBuf> {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply crates/aptos/src/common/types.rs:700]: The problem is that you are using dir already in `MovePackageDir`. So its natural to say e.g. `package_dir: MovePackageDir`. However, it then looks strange if you do `let path = package_dir.get_package_dir()` ...

I can undo this if you do not like it.

File: crates/aptos/src/common/types.rs [hunk 5]
```diff
@@ -689,7 +689,15 @@ pub struct MovePackageDir {
 }
 
 impl MovePackageDir {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply crates/aptos/src/common/types.rs:691]: Yes, that would be the alternative.

File: crates/aptos/src/move_tool/mod.rs [hunk 6]
```diff
@@ -280,6 +285,13 @@ pub struct PublishPackage {
     move_options: MovePackageDir,
     #[clap(flatten)]
     txn_options: TransactionOptions,
+    /// Whether to use the new publishing flow.
+    #[clap(long)]
+    new_flow: bool,
```
[the remark this answers is not in the document]
Comment (wrwg) [reply crates/aptos/src/move_tool/mod.rs:290]: There are more updates needed, and specifically, there are bugs in vm publishing which we first need to fix. For the next devnet release, we can look at next week whether we are ready to switch or hide this option.

File: crates/aptos/src/move_tool/mod.rs [hunk 7]
```diff
@@ -289,31 +301,42 @@ impl CliCommand<TransactionSummary> for PublishPackage {
...
+            upgrade_policy,
+        } = self;
+        let package = BuiltPackage::build(move_options, true, true)?;
+        let compiled_units = package.extract_code();
+        if !new_flow {
+            if upgrade_policy.is_some() {
+                return Err(CliError::CommandArgumentError(
+                    "`--upgrade-policy` can only be used with the `--new-publish` flow".to_owned(),
+                ));
+            }
```
[the remark this answers is not in the document]
Comment (wrwg) [reply crates/aptos/src/move_tool/mod.rs:317]: Fixed the old flag name. Yes clap can do this but too lazy to look up how ;-)

File: crates/aptos/src/move_tool/mod.rs (hunk 2 above)
Comment (wrwg) [reply crates/aptos/src/move_tool/mod.rs:332]: I agree with you conceptually, but this is really a hard panic, something which must never happen (and wont). I generally use the following approach for tools like this:

- Any error which is consequence of user input is gracefully reported
- Internal programming errors still crash

WDYT?

File: testsuite/smoke-test/src/aptos/move_test_helpers.rs (hunk 3 above)
Comment (wrwg) [reply testsuite/smoke-test/src/aptos/move_test_helpers.rs:44]: Same here

File: crates/aptos/debug-move-example/sources/DebugDemo.move [hunk 8]
```diff
@@ -1,7 +1,7 @@
 module DebugDemo::Message {
     use std::string;
     use std::signer;
-    use std::Debug;
```
Comment (areshand) [inline crates/aptos/debug-move-example/sources/DebugDemo.move:4]: should this be aptos_stdlib to be consistent with its package?

**Pull request merged**